crates/sui-core/src/unit_tests/gas_tests.rs

```diff
@@ -361,8 +360,9 @@ async fn test_publish_gas() -> anyhow::Result<()> {
 
     // Make sure that we are not charging storage cost at failure.
     assert_eq!(gas_cost.storage_cost, 0);
+    assert_eq!(gas_cost.storage_rebate, 0);
     // Upon failure, we should only be charging the expected computation cost.
-    assert_eq!(gas_cost.gas_used(), computation_cost);
```

Comment:
This test raises an interesting question: should you be able to run out of gas if the total gas used is less than your budget?

The answer to this question before this PR was "yes". The answer after this PR is "no".

Reply:
How does this work exactly? If you run out of gas, do we try to satisfy all storage cost first, and then use rest as the computation cost? Also curious what happens in the current impl if we run out of gas smashing coins.

Reply:
That assert has been changed as it was bogus (and in fact failed).
if we run OOG we simply charge the budget under all circumstances. The charges in computation and storage may match (should) but I do not think it should matter. The OOG error and the budget being charged are the only thing that should matter.
Smashing coins is a very good point and something we should test for and tune still.
The principle is that the minimal cost of a transaction would be enough to cover for gas operations up to the max amount of coins usable for gas (to be determined, now 256). Then we should go on charging for gas operations on entry and if we run out of gas we collect all the coins.
Transactions are rejected at signing if they are under min budget so hopefully that covers the issue.
Thoughts?